sdk/storage/azure-storage-common/test/ut/test_base.hpp

```diff
@@ -0,0 +1,211 @@
+          std::unique_ptr<Azure::Core::IO::BodyStream>& stream)
+      {
+        Azure::Core::Context context;
+        return stream->ReadToEnd(context);
+      }
+
+      inline std::vector<uint8_t> ReadBodyStream(
+          std::unique_ptr<Azure::Core::IO::BodyStream>&& stream)
+      {
+        return ReadBodyStream(stream);
+      }
+
+      std::vector<uint8_t> ReadFile(const std::string& filename);
+
+      void DeleteFile(const std::string& filename);
+
+      std::string InferSecondaryUrl(const std::string primaryUri);
+
+      inline std::string Base64EncodeText(const std::string& text)
+      {
+        return Azure::Core::Convert::Base64Encode(std::vector<uint8_t>(text.begin(), text.end()));
+      }
+
+      virtual void SetUp() override
+      {
+        Azure::Core::Test::TestBase::SetUpTestBase(AZURE_TEST_RECORDING_DIR);
+      }
+
+    public:
+      const static Azure::ETag DummyETag;
+      const static Azure::ETag DummyETag2;
+    };
+
+    constexpr inline unsigned long long operator""_KB(unsigned long long x) { return x * 1024; }
+    constexpr inline unsigned long long operator""_MB(unsigned long long x)
+    {
+      return x * 1024 * 1024;
+    }
+    constexpr inline unsigned long long operator""_GB(unsigned long long x)
+    {
+      return x * 1024 * 1024 * 1024;
+    }
+    constexpr inline unsigned long long operator""_TB(unsigned long long x)
+    {
+      return x * 1024 * 1024 * 1024 * 1024;
+    }
+
+    class CryptFunctionsTest : public StorageTest {
+    };
+
+    class ClientSecretCredentialTest : public StorageTest {
+
+    private:
+      std::unique_ptr<Azure::Storage::Blobs::BlobContainerClient> m_client;
+
+    protected:
+      std::shared_ptr<Core::Credentials::TokenCredential> m_credential;
+      std::string m_containerName;
+
+      // Required to rename the test propertly once the test is started.
+      // We can only know the test instance name until the test instance is run.
+      Azure::Storage::Blobs::BlobContainerClient const& GetClientForTest(
+          std::string const& testName)
+      {
+        // set the interceptor for the current test
+        m_testContext.RenameTest(testName);
+        return *m_client;
+      }
+
+      virtual void SetUp() override
+      {
+        StorageTest::SetUp();
+        m_containerName = Azure::Core::_internal::StringExtensions::ToLower(GetTestName());
+
+        m_credential = std::make_shared<Azure::Identity::ClientSecretCredential>(
+            AadTenantId(), AadClientId(), AadClientSecret());
+
+        Azure::Storage::Blobs::BlobClientOptions options;
+
+        m_client = InitTestClient<
+            Azure::Storage::Blobs::BlobContainerClient,
+            Azure::Storage::Blobs::BlobClientOptions>(
+            Azure::Storage::Blobs::BlobContainerClient::CreateFromConnectionString(
+                StandardStorageConnectionString(), m_containerName)
+                .GetUrl(),
+            &m_credential,
+            options);
```

Comment:
I don't think we need `InitTestClient` at all in the first place. What's the value of this function, with all the complexity of template stuff?

```C++
Azure::Storage::Blobs::BlobClientOptions options;
InitClientOptions(options);
m_client = std::make_unique<Azure::Storage::Blobs::BlobContainerClient>(
            Azure::Storage::Blobs::BlobContainerClient::CreateFromConnectionString(
                StandardStorageConnectionString(), m_containerName).GetUrl(),
            &m_credential,
            options);
```

Reply:
InitTestClient<ClientClass, OptionsClass>() make it possible to re-use this call from any service. It is useful for any test case that creates a client with a tokenCredential.

The method is not a must-call method, anyway. From the test case, we can call InitClientOptions and InitCredential as well.

At the end, what we need is to let the test-framework to take care of updating the client options and credentials to support playback.